File: tree/OC/LC/LC/LC105.m
//
//  LC105.m
//  LC
//
//  Created by ULS on 2/21/18.
//  Copyright © 2018 ULS. All rights reserved.
//

#import "LC105.h"
#import "TreeNode.h"

@interface LC105()

- (TreeNode *)buildTree:(int *)preorder preorderSize:(int)preorderSize inorder:(int *)inorder inorderSize:(int)inorderSize;

@end

@implementation LC105

- (TreeNode *)buildTree:(int *)preorder preorderSize:(int)preorderSize inorder:(int *)inorder inorderSize:(int)inorderSize {
    return [self dfs:preorder pstart:0 pend:preorderSize - 1 inorder:inorder istart:0 iend:inorderSize - 1];
}

- (TreeNode *)dfs:(int *)preorder pstart:(int)pstart pend:(int)pend inorder:(int *)inorder istart:(int)istart iend:(int)iend{
    if(pstart > pend || istart > iend) {
        return NULL;
    }
    
    TreeNode *root = [[TreeNode alloc] initWith:preorder[pstart]];
    int rootIndex = -1;
    for(int i = 0; i <= iend; i++) {
        if(root.val == inorder[i]){
            rootIndex = i;
            break;
        }
    }
    
    if(rootIndex == -1) {
        return root;
    }
    int offSet = rootIndex - istart;
    root.left = [self dfs:preorder pstart:pstart + 1 pend:pstart + offSet inorder:inorder istart:istart iend:rootIndex - 1];
    root.right = [self dfs:preorder pstart:pstart + offSet + 1 pend:pend inorder:inorder istart:rootIndex + 1 iend:iend];
    return root;
}

- (void)test {
    
}
@end

File: LCJava/src/com/leetcode/Tree/LC105.java
package com.leetcode.Tree;
import com.leetcode.Helper.*;


public class LC105 {

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return dfs(0, preorder.length - 1, 0, inorder.length - 1, preorder, inorder);
    }

    public TreeNode dfs(int ps,
                        int pe,
                        int is,
                        int ie,
                        int[] preorder,
                        int[] inorder) {
        if (ps > pe || is > ie) {
            return null;
        }
        TreeNode root = new TreeNode(preorder[ps]);

        int rootIdx = 0; // Index of current root in inorder
        for (int i = is; i <= ie; i++) {
            if (inorder[i] == root.val) {
                rootIdx = i;
            }
        }

        root.left = dfs(ps + 1, pe, is, rootIdx - 1, preorder, inorder);
        // rootIdx - is is the left tree length
        // ps + (rootIdx - is) cross the root and left tree.
        // ps + (rootIdx - is) + 1 is the right tree of preorder.

        root.right = dfs(ps + rootIdx - is + 1, pe, rootIdx + 1, ie, preorder, inorder);
        return root;
    }

    public static void main(String[] args) {
        LC105 obj = new LC105();
        TreeNode root = obj.buildTree(new int[]{3,9,20,15,7}, new int[]{9,3,15,20,7});
        // TreeUtils.printTree(root);
        TreeUtils.printTreeByLevel(root);
    }
}
